src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -232,96 +403,107 @@ void Foam::directAMI<SourcePatch, TargetPatch>::calculate
+    if (!this->srcPatch_.size())
     {
         return;
     }
+    else if (!this->tgtPatch_.size())
+    {
+        WarningIn
+            (
+                "void Foam::AMIMethod<SourcePatch, TargetPatch>::initialise"
+                    "("
+                    "labelListList&, "
+                    "scalarListList&, "
+                    "labelListList&, "
+                    "scalarListList&, "
+                    "label&, "
+                    "label&"
+                    ")"
+            )
+        << this->srcPatch_.size() << " source faces but no target faces" << endl;
 
+        return;
+    }
 
-    // temporary storage for addressing and weights
-    List<DynamicList<label> > srcAddr(this->srcPatch_.size());
-    List<DynamicList<label> > tgtAddr(this->tgtPatch_.size());
+    const pointgpuField& tgtPoints = this->tgtPatch_.getPoints();
+    const pointgpuField& srcPoints = this->srcPatch_.getPoints();
 
+    const faceDatagpuList& tgtFaces = this->tgtPatch_.getFaces();
+    const faceDatagpuList& srcFaces = this->srcPatch_.getFaces();
 
-    // construct weights and addressing
-    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    vectorgpuField srcCf(srcFaces.size());
 
-    // list of faces currently visited for srcFaceI to avoid multiple hits
-    DynamicList<label> srcSeeds(10);
+    thrust::transform
+    (
+        srcFaces.begin(),
+        srcFaces.end(),
+        srcCf.begin(),
+        faceCentreFunctor
+        (
+            this->srcPatch_.getFaceNodes().data(),
+            srcPoints.data()
+        )
+    );
 
-    // list to keep track of tgt faces used to seed src faces
-    labelList srcTgtSeed(srcAddr.size(), -1);
-    srcTgtSeed[srcFaceI] = tgtFaceI;
+    labelgpuList srcCount(this->srcPatch_.size(), 0);
+    labelgpuList tgtCount(this->tgtPatch_.size(), 0);
 
-    // list to keep track of whether src face can be mapped
-    // 1 = mapped, 0 = untested, -1 = cannot map
-    labelList mapFlag(srcAddr.size(), 0);
+    gpuList<labelList> srcAddr(this->srcPatch_.size(), labelList(15, -1));
+    gpuList<labelList> tgtAddr(this->tgtPatch_.size(), labelList(15, -1));
 
-    label nTested = 0;
-    DynamicList<label> nonOverlapFaces;
-    do
+    for (int i = 0; i < srcFaces.size(); ++i)
     {
-        srcAddr[srcFaceI].append(tgtFaceI);
-        tgtAddr[tgtFaceI].append(srcFaceI);
-
-        mapFlag[srcFaceI] = 1;
-
-        nTested++;
-
-        // Do advancing front starting from srcFaceI, tgtFaceI
-        appendToDirectSeeds
+        thrust::counting_iterator<unsigned> tgtStart(0);
+        thrust::counting_iterator<unsigned> tgtEnd = tgtStart + tgtFaces.size();
+        thrust::constant_iterator<unsigned> srcStart(i);
+        thrust::transform
         (
-            mapFlag,
-            srcTgtSeed,
-            srcSeeds,
-            nonOverlapFaces,
-            srcFaceI,
-            tgtFaceI
+            tgtStart,
+            tgtEnd,
+            srcStart,
+            ShootRayFunctor
```

Comment:
This call is wrong because the `ShootRayFunctor` is a binary functor. This needs to be fixed. Currently this gives an error when compiling.